File: parser/intent.go
package parser

import (
	"regexp"
	"strings"
)

var summaryIntentPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\bchi\s*tieu\s*thang\s*nay\b`),
	regexp.MustCompile(`\btong\s*chi\s*thang\s*nay\b`),
	regexp.MustCompile(`\bxem\s*chi\s*thang\s*nay\b`),
	regexp.MustCompile(`\bthong\s*ke\s*thang\s*nay\b`),
	regexp.MustCompile(`\bbao\s*cao\s*thang\s*nay\b`),
	regexp.MustCompile(`\bbao\s*cao\s*chi\s*tieu\b`),
	regexp.MustCompile(`\bchi\s*tieu\s*thang\s+(?:0?[1-9]|1[0-2])(?:\b|/)`),
	regexp.MustCompile(`\btong\s*chi\s*thang\s+(?:0?[1-9]|1[0-2])(?:\b|/)`),
	regexp.MustCompile(`\bxem\s*chi\s*thang\s+(?:0?[1-9]|1[0-2])(?:\b|/)`),
	regexp.MustCompile(`\bthong\s*ke\s*thang\s+(?:0?[1-9]|1[0-2])(?:\b|/)`),
	regexp.MustCompile(`\bbao\s*cao\s*thang\s+(?:0?[1-9]|1[0-2])(?:\b|/)`),
	regexp.MustCompile(`^/summary(?:\s|$)`),
}

func DetectMonthlySummaryIntent(input string) bool {
	normalized := normalizeForIntent(input)
	if normalized == "" {
		return false
	}
	for _, pattern := range summaryIntentPatterns {
		if pattern.MatchString(normalized) {
			return true
		}
	}
	return false
}

func normalizeForIntent(input string) string {
	input = strings.ToLower(strings.TrimSpace(input))
	var b strings.Builder
	space := false
	for _, r := range input {
		r = foldVietnameseRune(r)
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '/' {
			if space && b.Len() > 0 {
				b.WriteByte(' ')
			}
			space = false
			b.WriteRune(r)
			continue
		}
		space = true
	}
	return strings.TrimSpace(b.String())
}

func foldVietnameseRune(r rune) rune {
	switch r {
	case 'à', 'á', 'ạ', 'ả', 'ã', 'â', 'ầ', 'ấ', 'ậ', 'ẩ', 'ẫ', 'ă', 'ằ', 'ắ', 'ặ', 'ẳ', 'ẵ':
		return 'a'
	case 'è', 'é', 'ẹ', 'ẻ', 'ẽ', 'ê', 'ề', 'ế', 'ệ', 'ể', 'ễ':
		return 'e'
	case 'ì', 'í', 'ị', 'ỉ', 'ĩ':
		return 'i'
	case 'ò', 'ó', 'ọ', 'ỏ', 'õ', 'ô', 'ồ', 'ố', 'ộ', 'ổ', 'ỗ', 'ơ', 'ờ', 'ớ', 'ợ', 'ở', 'ỡ':
		return 'o'
	case 'ù', 'ú', 'ụ', 'ủ', 'ũ', 'ư', 'ừ', 'ứ', 'ự', 'ử', 'ữ':
		return 'u'
	case 'ỳ', 'ý', 'ỵ', 'ỷ', 'ỹ':
		return 'y'
	case 'đ':
		return 'd'
	default:
		return r
	}
}

File: src/parser/intent.ts
const SUMMARY_INTENT_PATTERNS = [
  /\bchi\s*tieu\s*thang\s*nay\b/u,
  /\btong\s*chi\s*thang\s*nay\b/u,
  /\bxem\s*chi\s*thang\s*nay\b/u,
  /\bthong\s*ke\s*thang\s*nay\b/u,
  /\bbao\s*cao\s*thang\s*nay\b/u,
  /\bbao\s*cao\s*chi\s*tieu\b/u,
  /\bchi\s*tieu\s+thang\s+(?:0?[1-9]|1[0-2])(?:\b|\/)/u,
  /\btong\s*chi\s+thang\s+(?:0?[1-9]|1[0-2])(?:\b|\/)/u,
  /\bxem\s*chi\s+thang\s+(?:0?[1-9]|1[0-2])(?:\b|\/)/u,
  /\bthong\s*ke\s+thang\s+(?:0?[1-9]|1[0-2])(?:\b|\/)/u,
  /\bbao\s*cao\s+thang\s+(?:0?[1-9]|1[0-2])(?:\b|\/)/u,
  /^\/summary(?:\s|$)/u,
];

export function detectMonthlySummaryIntent(input: string): boolean {
  const normalized = normalizeForIntent(input);
  return normalized !== '' && SUMMARY_INTENT_PATTERNS.some((pattern) => pattern.test(normalized));
}

export function normalizeForIntent(input: string): string {
  let normalized = '';
  let pendingSpace = false;
  for (const original of input.trim().toLowerCase()) {
    const char = foldVietnameseChar(original);
    if ((char >= 'a' && char <= 'z') || (char >= '0' && char <= '9') || char === '/') {
      if (pendingSpace && normalized) normalized += ' ';
      pendingSpace = false;
      normalized += char;
    } else {
      pendingSpace = true;
    }
  }
  return normalized.trim();
}

function foldVietnameseChar(char: string): string {
  const groups: Record<string, string> = {
    à: 'a',
    á: 'a',
    ạ: 'a',
    ả: 'a',
    ã: 'a',
    â: 'a',
    ầ: 'a',
    ấ: 'a',
    ậ: 'a',
    ẩ: 'a',
    ẫ: 'a',
    ă: 'a',
    ằ: 'a',
    ắ: 'a',
    ặ: 'a',
    ẳ: 'a',
    ẵ: 'a',
    è: 'e',
    é: 'e',
    ẹ: 'e',
    ẻ: 'e',
    ẽ: 'e',
    ê: 'e',
    ề: 'e',
    ế: 'e',
    ệ: 'e',
    ể: 'e',
    ễ: 'e',
    ì: 'i',
    í: 'i',
    ị: 'i',
    ỉ: 'i',
    ĩ: 'i',
    ò: 'o',
    ó: 'o',
    ọ: 'o',
    ỏ: 'o',
    õ: 'o',
    ô: 'o',
    ồ: 'o',
    ố: 'o',
    ộ: 'o',
    ổ: 'o',
    ỗ: 'o',
    ơ: 'o',
    ờ: 'o',
    ớ: 'o',
    ợ: 'o',
    ở: 'o',
    ỡ: 'o',
    ù: 'u',
    ú: 'u',
    ụ: 'u',
    ủ: 'u',
    ũ: 'u',
    ư: 'u',
    ừ: 'u',
    ứ: 'u',
    ự: 'u',
    ử: 'u',
    ữ: 'u',
    ỳ: 'y',
    ý: 'y',
    ỵ: 'y',
    ỷ: 'y',
    ỹ: 'y',
    đ: 'd',
  };
  return groups[char] ?? char;
}
